keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+         * [Possible Extensions](#possible-extensions)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+This proposal aims at allowing Pod resource requests & limits to be updated
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally, PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not sufficient,
+* load has decreased significantly and allocated resources are unused and wasted,
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
```

Comment:
Also, how are you getting the allocated resources from the runtime?

Reply:
We are not getting this from runtime. When new Pod and its container are successfully started, we set this value to that in PodSpec.Container.Resources. For resize, we set it to the desired value upon successful completion of UpdateContainerResources API. I realize that for memory decrease, we may have to use force reclaim loop that Derek has been referring to.

Reply:
What's the source of truth if kubelet restarts? It has to re-discover this information from somewhere (and not from the kubernetes apiserver)

Reply:
Good point. I learned recently from David's pointers to API conventions that PodStatus should be reconstructable from observations, which is not the case if KL were to restart. One possible solution may be to move ResourcesAllocated to PodSpec that is written to by Kubelet via a subresource. Does that sound plausible?

Reply:
This belongs to status; moving this from status to spec doesn't solve the fundamental issue of kubelet not being able to detect the true allocated resources.

Kubelet could crash before writing this information back to the apiserver for various reasons. We usually rely on information discoverable locally (i.e., source of truth) for this kind information.

Reply:
I don't think per container memory allocation is stored anywhere on the Node today (and it should not be). Limits are discoverable for pod and docker containers (maybe CRI can be extended to return cgroup info), but requests (current usage) are not discoverable.

Having the allocation information in the spec does provide a trusted source for the node to work with. In our scenario, the ResourceAllocations either match Resources (desired) or not. If kubelet quits before after a pod and its containers have been resized, it will work on it upon restart as if it were handling a new resize request. If it quits after pod, containers have been resized, but before it has written ResourceAllocations to spec, then it will repeat the resize on restart treating it as a new request. When it adds the Pods, it will apply the same resize admission logic, and it should succeed just as it did before crash. It will call UpdateContainerResources which merely writes already updated cpu, memory limit values. So this will work.

Do you have a better suggestion considering we cannot use PodStatus as a store?

Reply:
And I do agree this belongs in PodStatus. I looked through it some more, but I currently don't see a way short of storing metadata with the containers about current resource allocations. Currently when kubelet restarts, it drives to  bring the current Pod states discovered on the node equal to what the apiserver tells it, and it is storing node state in that sense. Do you think adding metadata is a good idea? Or is there some other way we can store and get this info locally?

Reply:
We added ResourcesAllocated to PodSpec that resolves the persistence of current allocations. Resolving this issue.